In [1]:
#Gerekli kütüphaneleri import edelim.

import pandas as pd
import numpy as np
import tensorflow as tf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

In [2]:
import random 
tf.random.set_seed(0)
np.random.seed(0)
random.seed(0)

In [3]:
#veri setini okuyalım
df=pd.read_csv("C:/Users/Asus/.cache/kagglehub/datasets/uciml/iris/versions/2/Iris.csv")
df

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
0      1            5.1           3.5            1.4           0.2   
1      2            4.9           3.0            1.4           0.2   
2      3            4.7           3.2            1.3           0.2   
3      4            4.6           3.1            1.5           0.2   
4      5            5.0           3.6            1.4           0.2   
..   ...            ...           ...            ...           ...   
145  146            6.7           3.0            5.2           2.3   
146  147            6.3           2.5            5.0           1.9   
147  148            6.5           3.0            5.2           2.0   
148  149            6.2           3.4            5.4           2.3   
149  150            5.9           3.0            5.1           1.8   

            Species  
0       Iris-setosa  
1       Iris-setosa  
2       Iris-setosa  
3       Iris-setosa  
4       Iris-setosa  
..              ...  
145  Iris-virginica  
146  Iris-virginica  
147  Iris-virginica  
148  Iris-virginica  
149  Iris-virginica  

[150 rows x 6 columns]

In [4]:
#Özet istatistikler.

print(df.describe())

               Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
count  150.000000     150.000000    150.000000     150.000000    150.000000
mean    75.500000       5.843333      3.054000       3.758667      1.198667
std     43.445368       0.828066      0.433594       1.764420      0.763161
min      1.000000       4.300000      2.000000       1.000000      0.100000
25%     38.250000       5.100000      2.800000       1.600000      0.300000
50%     75.500000       5.800000      3.000000       4.350000      1.300000
75%    112.750000       6.400000      3.300000       5.100000      1.800000
max    150.000000       7.900000      4.400000       6.900000      2.500000


In [5]:
X = df.drop(columns=["Id", "Species"])
y = LabelEncoder().fit_transform(df["Species"])
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

Bu işlemde LabelEncoder, çiçek türlerini şu şekilde kodlar:

Iris-setosa: 0
Iris-versicolor: 1
Iris-virginica: 2

In [6]:
model = tf.keras.Sequential([
    #8 nörondan oluşan ve ReLU aktivasyon fonksiyonu kullanan bir katman. 
    # Giriş şekli, eğitim verisinin özellik sayısına göre belirleniyor.
    tf.keras.layers.Dense(8, activation='relu', input_shape=(X_train.shape[1],)),
    #3 nöronlu bir çıktı katmanı olup, sınıflandırma yapabilmek için softmax aktivasyon fonksiyonu kullanır.
    tf.keras.layers.Dense(3, activation='softmax')
])

#Model, Adam optimizasyonu ve sparse_categorical_crossentropy kayıp fonksiyonu ile derleniyor.
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

c:\Users\Asus\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [7]:
#Her epoch, modelin tüm eğitim verisi üzerinde bir eğitim adımı gerçekleştirdiği bir döngüdür.

model.fit(X_train, y_train, epochs=50, validation_data=(X_test, y_test))

Epoch 1/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 1s 80ms/step - accuracy: 0.3081 - loss: 5.6739 - val_accuracy: 0.3667 - val_loss: 4.2875
Epoch 2/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step - accuracy: 0.3081 - loss: 5.4977 - val_accuracy: 0.3667 - val_loss: 4.1337
Epoch 3/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.3081 - loss: 5.3247 - val_accuracy: 0.3667 - val_loss: 3.9823
Epoch 4/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.3081 - loss: 5.1549 - val_accuracy: 0.3667 - val_loss: 3.8332
Epoch 5/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step - accuracy: 0.3081 - loss: 4.9883 - val_accuracy: 0.3667 - val_loss: 3.6863
Epoch 6/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step - accuracy: 0.3081 - loss: 4.8249 - val_accuracy: 0.3667 - val_loss: 3.5416
Epoch 7/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step - accuracy: 0.3081 - loss: 4.6646 - val_accuracy: 0.3667 - val_loss: 3.3994
Epoch 8/50
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step - accuracy: 0.3081 - loss: 4.5078 - val_accuracy: 0.3667 - val_loss: 3.2615


In [8]:
#etiketlerin nasıl verildiğini görelim.
print(LabelEncoder().fit(df["Species"]).classes_)

['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']


In [9]:
x_new = np.array([[4.6, 3.1, 1.5, 0.2]])
prediction = model.predict(x_new)
print("Prediction:", np.argmax(prediction))
print("Prediction probabilities:", prediction)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 60ms/step
Prediction: 0
Prediction probabilities: [[0.64121383 0.22735171 0.13143447]]


Prediction probabilities: [[0.64121383 0.22735171 0.13143447]]: Model, her bir sınıfa olan olasılıkları tahmin eder.

Iris-setosa için olasılık: 0.641 (yaklaşık %64)
Iris-versicolor için olasılık: 0.227 (yaklaşık %23)
Iris-virginica için olasılık: 0.131 (yaklaşık %13)
Model, en yüksek olasılığı seçerek Iris-setosa sınıfını tahmin etmiştir.